File: src/services/gemScraper.js
const axios = require('axios');
const Tender = require('../models/Tender');

const scrape_gem_tenders = async (maxPages = 5) => {
  try {
    console.log('Loading GEM bids page...');
    const pageResp = await axios.get('https://bidplus.gem.gov.in/all-bids');
    const html = pageResp.data;

    const match = html.match(/id="chash"\s+type="hidden"\s+value="([^"]+)"/);
    if (!match) {
      console.log('Could not find CSRF token');
      return 0;
    }

    const csrfToken = match[1];
    console.log(`CSRF token: ${csrfToken.substring(0, 20)}...`);

    let count = 0;

    for (let pageNum = 1; pageNum <= maxPages; pageNum++) {
      console.log(`\nFetching page ${pageNum}...`);

      const payload = {
        page: pageNum,
        param: { searchBid: '', searchType: 'fullText' },
        filter: {
          bidStatusType: 'ongoing_bids',
          byType: 'all',
          highBidValue: '',
          byEndDate: { from: '', to: '' },
          sort: 'Bid-End-Date-Latest'
        }
      };

      try {
        const resp = await axios.post('https://bidplus.gem.gov.in/all-bids-data',
          `payload=${encodeURIComponent(JSON.stringify(payload))}&csrf_bd_gem_nk=${csrfToken}`,
          { headers: { 'Content-Type': 'application/x-www-form-urlencoded' } }
        );

        const docs = resp.data?.response?.response?.docs || [];
        if (!docs.length) {
          console.log(`No more tenders on page ${pageNum}. Stopping.`);
          break;
        }

        for (const doc of docs) {
          try {
            const extractStr = (val, defaultVal = 'N/A') => {
              if (Array.isArray(val)) return val.length ? val[0] : defaultVal;
              return val ? String(val) : defaultVal;
            };

            let title = extractStr(doc.bbt_title);
            if (title === 'N/A') title = extractStr(doc.b_category_name);

            const ministry = extractStr(doc.ba_official_details_minName, '');
            const dept = extractStr(doc.ba_official_details_deptName, '');
            const department = (ministry && dept) ? `${ministry} - ${dept}` : ministry || dept || 'N/A';

            const bidId = extractStr(doc.b_id, '');
            const bidNumber = extractStr(doc.b_bid_number, '');
            const link = `https://bidplus.gem.gov.in/showbidDocument/${bidId}`;

            const existing = await Tender.findOne({ reference_link: link });
            if (existing) continue;

            const category = extractStr(doc.b_category_name, 'General');

            await Tender.create({
              title,
              department,
              closing_date: extractStr(doc.final_end_date_sort),
              published_date: extractStr(doc.final_start_date_sort),
              tender_value: 'N/A',
              reference_link: link,
              summary: `Bid: ${bidNumber} | ${title}`,
              location: 'India',
              source: 'gem',
              category,
              state: 'All India'
            });

            count++;
            console.log(`Saved: ${title.substring(0, 60)}...`);
          } catch (err) {
            if (err.code === 11000) continue;
            console.log(`Error parsing bid: ${err.message}`);
          }
        }
      } catch (err) {
        console.log(`Error fetching page ${pageNum}: ${err.message}`);
        break;
      }

      await new Promise(r => setTimeout(r, 1000));
    }

    console.log(`\nTotal GEM tenders saved: ${count}`);
    return count;
  } catch (err) {
    console.error('GEM scraper error:', err.message);
    return 0;
  }
};

module.exports = { scrape_gem_tenders };

File: gem_scraper.py
import httpx
import re
from crud import create_tender, get_tender_by_reference_link
from models import TenderCreate


async def scrape_gem_tenders(max_pages: int = 5):
    """Scrape GEM tenders from bidplus.gem.gov.in via AJAX API."""
    async with httpx.AsyncClient() as client:
        print("Loading GEM bids page...")
        resp = await client.get("https://bidplus.gem.gov.in/all-bids")

        match = re.search(r'id="chash"\s+type="hidden"\s+value="([^"]+)"', resp.text)
        if not match:
            print("Could not find CSRF token")
            return 0

        csrf_token = match.group(1)
        print(f"CSRF token: {csrf_token[:20]}...")

        count = 0
        for page_num in range(1, max_pages + 1):
            print(f"\nFetching page {page_num}...")

            payload = {
                "page": page_num,
                "param": {"searchBid": "", "searchType": "fullText"},
                "filter": {
                    "bidStatusType": "ongoing_bids",
                    "byType": "all",
                    "highBidValue": "",
                    "byEndDate": {"from": "", "to": ""},
                    "sort": "Bid-End-Date-Latest"
                }
            }

            try:
                resp = await client.post(
                    "https://bidplus.gem.gov.in/all-bids-data",
                    data={"payload": str(payload), "csrf_bd_gem_nk": csrf_token}
                )
                data = resp.json()
                docs = data.get("response", {}).get("response", {}).get("docs", [])

                if not docs:
                    print(f"No more tenders on page {page_num}. Stopping.")
                    break

                for doc in docs:
                    try:
                        def extract_str(val, default="N/A"):
                            if isinstance(val, list):
                                return val[0] if val else default
                            return str(val) if val else default

                        title = extract_str(doc.get("bbt_title", ""))
                        if title == "N/A":
                            title = extract_str(doc.get("b_category_name", ["N/A"]))

                        ministry = extract_str(doc.get("ba_official_details_minName", ""), "")
                        dept = extract_str(doc.get("ba_official_details_deptName", ""), "")
                        department = f"{ministry} - {dept}" if ministry and dept else ministry or dept or "N/A"

                        bid_id = extract_str(doc.get("b_id", ""), "")
                        bid_number = extract_str(doc.get("b_bid_number", ""), "")

                        link = f"https://bidplus.gem.gov.in/showbidDocument/{bid_id}"

                        existing = await get_tender_by_reference_link(link)
                        if existing:
                            continue

                        tender = TenderCreate(
                            title=title,
                            department=department,
                            closing_date=extract_str(doc.get("final_end_date_sort", "")),
                            published_date=extract_str(doc.get("final_start_date_sort", "")),
                            tender_value="N/A",
                            reference_link=link,
                            summary=f"Bid: {bid_number} | {title}",
                            location="India",
                            source="gem",
                            category=extract_str(doc.get("b_category_name", ["General"])),
                            state="All India"
                        )
                        await create_tender(tender)
                        count += 1
                        print(f"Saved: {tender.title[:60]}...")
                    except Exception as e:
                        print(f"Error parsing bid: {e}")
                        continue

            except Exception as e:
                print(f"Error fetching page {page_num}: {e}")
                break

        print(f"\nTotal GEM tenders saved: {count}")
        return count


if __name__ == "__main__":
    import asyncio
    asyncio.run(scrape_gem_tenders())